### ISSUE TITLE
Non-ASCII characters in sample PRD and REF

### ISSUE THREAD

**Issue by Shujian2015** [freewym/espresso#11]
Hi,
    During training swbd recipe, the log on screen:

> | sample PRD: \xe2\x96\x81maybe\xe2\x96\x81c'
| sample REF: b'\xe2\x96\x81[vocalized-noise]'

Also the WER on swbd val set is very large. Is this normal? Thanks in advance.

**Comment by freewym** [comment]
Did you check if `text` file is displayed as normal

**Comment by Shujian2015** [comment]
Hi @freewym ,
`text` looks normal:

> sw02054-A_000204-000790 so let me tell you a little bit about the weather here last sunday night we had an ice storm

but `token_text` looks like this:

> sw02054-A_000204-000790 b'\xe2\x96\x81so \xe2\x96\x81let \xe2\x96\x81me \xe2\x96\x81tell \xe2\x96\x81you \xe2\x96\x81a \xe2\x96\x81little \xe2\x96\x81bit \xe2\x96\x81about \xe2\x96\x81the \xe2\x96\x81weather \xe2\x96\x81here \xe2\x96\x81last \xe2\x96\x81su n d a y \xe2\x96\x81night \xe2\x96\x81we \xe2\x96\x81had \xe2\x96\x81an \xe2\x96\x81i ce \xe2\x96\x81st or m'

**Comment by freewym** [comment]
I think you need to check each output file of sentencepiece related code in stage 2, to see where it started to be abnormal.

**Comment by Shujian2015** [comment]
Thanks @freewym !

**Comment by Shujian2015** [comment]
Hi @freewym , I have another question about swbd recipe. I am running it with 8 GPUs with `update-freq` to be 8. Now ~90% of the training time is purely CPU calculation. CPU usage is almost 100% on 64 cores. Do you know what takes it so long? Thanks in advance.

**Comment by freewym** [comment]
I am not sure. Maybe you can try the LM training task with the same number of GPUs and update-freq, to see if the CPU time still dominates. This can help identify whether is is from Espresso code or original fairseq.

**Comment by Shujian2015** [comment]
Thanks @freewym . The LM part works fine: I use update-freq = 8 and only ~ 8 CPU cores are fully used. GPUs are busy most of the times.

**Comment by Shujian2015** [comment]
Removing `--ddp-backend no_c10d` in this line seems to solve my issue:
https://github.com/freewym/espresso/blob/master/examples/asr_swbd/run.sh#L252

**Comment by freewym** [comment]
Does removing no_c10d help? I added it because scheduled sampling will crash without it.

**Comment by Shujian2015** [comment]
no_c10d doesn't help.

**Comment by freewym** [comment]
Can you set on `--disable-validation` so that there is no validation requiring sync between GPUs, and set a very large `update-freq` so that there will not even be a sync within one epoch, and then start the training to see what will happen.

**Comment by Shujian2015** [comment]
I added `--update-freq 100000 --disable-validation`. 
Now all 96 CPU cores have been busy for a long time. GPUs are all zero usage and ~14% of memory is occupied.

**Comment by freewym** [comment]
Does it suggest the more update-freq, the more CPU usage? Can you use cProfile to find out which function(s) takes long time?

**Comment by Shujian2015** [comment]
It looks so.

**Comment by freewym** [comment]
...And what about using 8 GPUs and  update-freq=1?

**Comment by Shujian2015** [comment]
Hi @freewym, with update-freq=1, the training looks normal but one epoch takes about the same time with 2 GPUs. I don’t see any speed up.

**Comment by freewym** [comment]
Does using 1 or 2 GPUs with a very large update-freq also take long CPU time?

**Comment by Shujian2015** [comment]
Here is time for each epoch in different settings:

  | update-freq=1 | update-freq=8
-- | -- | --
2 GPU | wall 2331 \| train_wall 1620 | wall 2072 \| train_wall 1440
8 GPU | wall 3693 \| train_wall 566 | wall 3496 \| train_wall 442

For more GPUs, small update frequency seems to lead too much communication cost in gradient update while large update frequency lead to crazy data loading traffic jam.

**Comment by freewym** [comment]
How do you know the bottleneck is data loading with large upadte-freq? By data loading do you mean loading from disk? If it is, see if increasing [cache_size](https://github.com/freewym/espresso/blob/master/fairseq/data/scp_dataset.py#L79) could help.

**Comment by Shujian2015** [comment]
Sorry Just my feeling. With large update-freq, 90% of the time CPU usage is high while GPU is free. I don't quite clearly what it is doing there.

**Comment by freewym** [comment]
See if a larger cache_size can help. I don't have any other clues.

**Comment by Shujian2015** [comment]
Thanks @freewym , should I change that here? https://github.com/freewym/espresso/blob/master/fairseq/data/scp_dataset.py#L79

**Comment by freewym** [comment]
Yes

**Comment by Shujian2015** [comment]
Seems there is no influence from this parameter and I didn't see ScpCachedDataset get used in the codebase.

**Comment by freewym** [comment]
It is being used [here](https://github.com/freewym/espresso/blob/master/fairseq/tasks/speech_recognition.py#L179)

**Comment by Shujian2015** [comment]
Oh. Thanks.

**Comment by freewym** [comment]
Sorry I am unable to get 4 GPUs to test this case. Maybe you can also try fairseq's ASR recipe to see if it works in your setup.

**Comment by Shujian2015** [comment]
I can try that next week. Thanks for all the help!

**Comment by Shujian2015** [comment]
I am facing the same issue when using FAIRSEQ example. I guess it is due to some issues with our server or my setup. I am closing this issue. I will comment if I can find a solution in the future.

**Comment by Shujian2015** [comment]
Update: Solved the issue with this code piece (and `--num-worker 1`): https://github.com/pytorch/fairseq/issues/1171#issuecomment-534981208

This seems a common issue to multiple V100 GPUs in cloud: 

- Some one has similar issue with Fairseq in AWS: https://github.com/pytorch/fairseq/issues/850
- Another issue: https://github.com/pytorch/fairseq/issues/564

**Comment by freewym** [comment]
Are you saying that, setting `--num-worker 1` is expected to reduce the CPU time, but have the shared memory issue, which in turns can be resolved by disabling shared memory as suggested in [this post](https://github.com/pytorch/fairseq/issues/1171#issuecomment-534981208)?

**Comment by Shujian2015** [comment]
Yeah.

**Comment by freewym** [comment]
Cool! What is the current CPU/GPU time ratio, and the total speedup compared with using single GPU?

**Comment by Shujian2015** [comment]
Now GPUs are mostly busy. 
2 GPUs with `update-freq`=1 take ~ 2000 sec per epoch. 
Now 6 GPUs with `update-freq`=6 take ~ 1000 sec per epoch (used to be ~ 3500 sec before). 

I really appreciated all the help @freewym !

**Comment by freewym** [comment]
So you are effectively increasing the batch size 18X (3 x 6) with the benefit of 2X speedup. would it be significantly slower when update-freq=2 or so?
You are welcome. You also helped me identify a potential problem.

**Comment by Shujian2015** [comment]
6 GPUs:
`--update-freq 6`: wall 1089 | train_wall 608
`--update-freq 2`: wall 1174 | train_wall 709

**Comment by freewym** [comment]
OK. It looks like a small update-freq  is still tolerable

**Comment by Shujian2015** [comment]
The '\xe2\x96\x81' issue seems only happen in python version lower than 3.7 (I am using 3.6).
And easy fix is to replace [this line ](https://github.com/freewym/espresso/blob/master/fairseq/data/data_utils.py#L236)with:
`sentence = sentence.replace(' ', '').replace('\\xe2\\x96\\x81', ' ').replace('\u2581', ' ').strip()`

**Comment by freewym** [comment]
Why could that happen? I remember it happened during spm_train.py or spm_encode.py in your case? How did you fix that?

**Comment by Shujian2015** [comment]
I am also confused but the method I mentioned above can fix it...

**Comment by freewym** [comment]
As this problem looks like purely from fairseq code, I'd rather not fix it in Espresso, but instead make a PR/issue directly to fairseq.

**Comment by Shujian2015** [comment]
For now I only fixed it in our internal fork. Unfortunately our server does't allow python 3.7 and I cannot make sure this is an issue for everyone or most of people.

**Comment by freewym** [comment]
Can you call sentencepiece from your python interactive mode, using the code like

https://github.com/pytorch/fairseq/blob/master/scripts/spm_train.py#L16 

to see if it is because of the original sentencepiece package or the fairseq's wrapper?

**Comment by Shujian2015** [comment]
Unfortunately my job duty recently moved from asr to search and I couldn't test anything in the near future.
